// @gomomento/sdk-core/dist/src/auth/tokens/token-scopes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.topicPublishSubscribe = exports.topicSubscribeOnly = exports.cacheReadOnly = exports.cacheReadWrite = void 0;
const token_scope_1 = require("./token-scope");
function cacheReadWrite(cacheSelector) {
    return {
        permissions: [{ role: token_scope_1.CacheRole.ReadWrite, cache: cacheSelector }],
    };
}
exports.cacheReadWrite = cacheReadWrite;
function cacheReadOnly(cacheSelector) {
    return {
        permissions: [{ role: token_scope_1.CacheRole.ReadOnly, cache: cacheSelector }],
    };
}
exports.cacheReadOnly = cacheReadOnly;
function topicSubscribeOnly(cacheSelector, topicSelector) {
    return {
        permissions: [
            {
                role: token_scope_1.TopicRole.SubscribeOnly,
                cache: cacheSelector,
                topic: topicSelector,
            },
        ],
    };
}
exports.topicSubscribeOnly = topicSubscribeOnly;
function topicPublishSubscribe(cacheSelector, topicSelector) {
    return {
        permissions: [
            {
                role: token_scope_1.TopicRole.PublishSubscribe,
                cache: cacheSelector,
                topic: topicSelector,
            },
        ],
    };
}
exports.topicPublishSubscribe = topicPublishSubscribe;
